Napatech: fixed SCTIME_ADD_USECS from util-time.h to allow correct pa…

ilya-bakhtin opened the pull request:
…cket timestamp calculation

Make sure these boxes are signed before submitting your Pull Request -- thank you.

- [x] I have read the contributing guide lines at https://docs.suricata.io/en/latest/devguide/codebase/contributing/contribution-process.html
- [x] I have signed the Open Information Security Foundation contribution agreement at https://suricata.io/about/contribution-agreement/
- [] I have updated the user guide (in doc/userguide/) to reflect the changes made (if applicable)

Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues) ticket:[6372](https://redmine.openinfosecfoundation.org/issues/6372)

Describe changes:
The SCTIME_ADD_USECS code is updated and timestamp timestamp-related code in source-napatech is fixed.
It seems the software was never tested with real hardware. Current 7.0.1 and master versions generate timestamps with seconds == 0.
The SV test is impossible since the code is only reachable when the card is available.
Technically it's possible to prepare unit test if it's required.

### Provide values to any of the below to override the defaults.

To use a pull request use a branch name like `pr/N` where `N` is the
pull request number.

Alternatively, `SV_BRANCH` may also be a link to an
OISF/suricata-verify pull-request.

```
SV_REPO=
SV_BRANCH=
SU_REPO=
SU_BRANCH=
LIBHTP_REPO=
LIBHTP_BRANCH=
```

jlucovsky commented:
Please split your commits -- the `util-time.c` changes to round up microseconds, when needed, are distinct from the Napatech time changes.

jlucovsky commented:
@ralpheastwood Could you review these changes?

ralpheastwood commented:
Hi @jlucovsky,  I've reviewed the changes and it looks correct to me.  I checked the original commit my colleague made and it was correct (and functionally the same as this commit) so it looks like I introduced the typo when I reworked the commits.

The change to `SCTIME_ADD_USECS` looks good to me too.

jlucovsky commented:
> Hi @jlucovsky, I've reviewed the changes and it looks correct to me. I checked the original commit my colleague made and it was correct (and functionally the same as this commit) so it looks like I introduced the typo when I reworked the commits.
> 
> The change to `SCTIME_ADD_USECS` looks good to me too.

Thanks for validating. I'm seeing a few anomalies with a timestamp (converted to it's display format): `"ts":"1970-01-01T00:00:00.796239Z"`

I'm not sure what this is yet -- I'll keep looking.

ralpheastwood commented:
For reference how are you testing and which time format are you using?

victorjulien commented:
> Hi @jlucovsky, I've reviewed the changes and it looks correct to me. I checked the original commit my colleague made and it was correct (and functionally the same as this commit) so it looks like I introduced the typo when I reworked the commits.

For future work, can you make sure to test the exact code you're submitting?

jlucovsky commented:
> For reference how are you testing and which time format are you using?

I'm using `NT_TIMESTAMP_TYPE_NATIVE_UNIX`; suricata's analyzing traffic that's being replayed and directed to the napatech input ports. I'm using an ET Open rules file.

jlucovsky commented:
> > Hi @jlucovsky, I've reviewed the changes and it looks correct to me. I checked the original commit my colleague made and it was correct (and functionally the same as this commit) so it looks like I introduced the typo when I reworked the commits.
> > The change to `SCTIME_ADD_USECS` looks good to me too.
> 
> Thanks for validating. I'm seeing a few anomalies with a timestamp (converted to it's display format): `"ts":"1970-01-01T00:00:00.796239Z"`
> 
> I'm not sure what this is yet -- I'll keep looking.

I'm not able to recreate the issue I saw with dates being close to the epoch.

jlucovsky commented:
I'd prefer that the `inline` function be changed to a macro,
e.g (but with proper formatting)
```
#define SCTIME_ADD_USECS(ts,us) \
    (SCTime_t)                                                                                     \
{\
     .secs = ts.secs + (ts.usecs + us) / 1000000, .usecs = (ts.usecs+us) % 1000000 \
}
```

ilya-bakhtin commented:
I'm sorry for the delay with answers, but today was a public holiday in the Czech Republic and I was not online.
I plan to make some polish tomorrow.
Now i'm thinking that it could be better to introduce a macro FROM_SECS_AND_USECS with two parameters.

ilya-bakhtin commented:
> I'm seeing a few anomalies with a timestamp (converted to it's display format): `"ts":"1970-01-01T00:00:00.796239Z"`

Current implementation results in these timestamps all the time.
And finally, it results in data race and corresponding segfaults since ref-counting is replaced by timestamps analysis in suri7.
As i can see correct timestamp values are crucial for normal flow lifetime analysis since ref counts are dropped in suri7.

ilya-bakhtin commented:
I've replaced inline static function by a macro.
I did not introduce FROM_SECS_AND_USECS mentioned above since napatech code applies a rounding to usecond that may result in its 1000000 value. So let's use SCTIME_ADD_USECS tat can handle this condition.

util-time.h and source-napatech.c are now updated by separate commits

jlucovsky commented on src/util-time.h:
```diff
@@ -57,7 +57,11 @@ typedef struct {
 #define SCTIME_MSECS(t)          (SCTIME_SECS(t) * 1000 + SCTIME_USECS(t) / 1000)
 #define SCTIME_ADD_SECS(ts, s)   SCTIME_FROM_SECS((ts).secs + (s))
-#define SCTIME_ADD_USECS(ts, us) SCTIME_FROM_USECS((ts).usecs + (us))
+#define SCTIME_ADD_USECS(ts, us)                                                                   \
```
In the macro, please change `us` to `(us)` (every where it's used).

jlucovsky commented:
> I'd prefer that the `inline` function be changed to a macro, e.g (but with proper formatting)
> 
> ```
> #define SCTIME_ADD_USECS(ts,us) \
>     (SCTime_t)                                                                                     \
> {\
>      .secs = ts.secs + (ts.usecs + us) / 1000000, .usecs = (ts.usecs+us) % 1000000 \
> }
> ```

Minor update: `s/us/(us)` in the macro implementation.

victorjulien commented:
> > I'd prefer that the `inline` function be changed to a macro, e.g (but with proper formatting)
> > ```
> > #define SCTIME_ADD_USECS(ts,us) \
> >     (SCTime_t)                                                                                     \
> > {\
> >      .secs = ts.secs + (ts.usecs + us) / 1000000, .usecs = (ts.usecs+us) % 1000000 \
> > }
> > ```
> 
> Minor update: `s/us/(us)` in the macro implementation.

Same for `ts`. Just always for all macros.

ilya-bakhtin commented:
yes sure.
my bad.
updated.

jlucovsky commented:
Replaces https://github.com/OISF/suricata/pull/9527

victorjulien commented:
I would like to see a new PR w/o force pushes, and that follows the commit message guidelines. Thanks!

ilya-bakhtin commented:
new https://github.com/OISF/suricata/pull/9833 is created

catenacyber commented:
So closing as replaced by https://github.com/OISF/suricata/pull/9833
